lib/CRMS/RightsPredictor.pm

```diff
@@ -0,0 +1,140 @@
+
+  if ($year !~ m/^-?\d{1,4}$/) {
+    return { year => undef, desc => "unsupported date format '$year'" };
+  }
+  my $country = $self->{record}->country;
+  if ($country eq 'United Kingdom' || $country eq 'Canada' || $country eq 'Australia') {
+    my $term = 70;
+    if ($is_crown) {
+      $term = 50;
+    } else {
+      if ($country eq 'Canada') {
+        $term = 50 if $year < 1972;
+      } elsif ($country eq 'Australia') {
+        $term = 50 if $year < 1955;
+      }
+    }
+    my $last_domestic_copyright = $year + $term;
+    return { year => $last_domestic_copyright,
+      desc => "Last $country © $last_domestic_copyright ($year + $term-year term)" };
+  }
+  #elsif ($country eq 'Spain') {
+  #  return { year => $year + 80, desc => "$country $year + 80-year term" };
+  #}
+  return { year => undef, desc => "country $country unknown term" };
+}
+
+# CRMS World specific. Predict best radio button (rights combo)
+# to choose based on user selections.
+# Returns data structure with the following fields (all but desc may be absent on error):
+# attr: rights attribute name
+# reason: rights reason name
+# rights: attr/reason string
+# last_domestic_copyright: year
+# desc: last_domestic_copyright desc plus more math, or error message
+sub rights {
+  my $self     = shift;
+  my $year     = shift || ''; # ADD or pub date entered by user
+  my $is_pub   = shift; # Treat $year as pub date for corporate authors
+  my $is_crown = shift; # Crown copyright found
+  #my $pub_year = shift; # Actual pub date when date range, entered by user
+  my $now      = shift || POSIX::strftime("%Y", localtime); # The "current" year
+
+  my $ldc = $self->last_domestic_copyright($year, $is_crown);
+  my $last_domestic_copyright = $ldc->{year};
+  return $ldc unless defined $last_domestic_copyright;
+  my $pub_year;
+  if ($is_pub) {
+    $pub_year = $year;
+  } else {
+    # Verify that we can reason about the year of publication:
+    # If there is a pub date range then it should not span the so-called 1023 boundary
+    $pub_year = $self->{record}->formatPubDate || '';
+    # We currently don't support date ranges at all.
+    # If we collect actual pub dates, we can sanity check that
+    # and then use it in place of the info in the catalog record 008.
+    if ($pub_year !~ m/^-?\d+$/) {
+      return { desc => "unsupported date format '$pub_year'" };
+    }
+  }
+  my ($attr, $reason);
+  my @desc = ( $ldc->{desc} );
+  if ($last_domestic_copyright < $now) {
+    push @desc, "$last_domestic_copyright < $now";
+    my $pub_plus_95 = $pub_year + 95;
+    # ICUS/GATT can still be applied to 1928 this year; next year it should not.
+    if ($last_domestic_copyright >= 1996 && $pub_plus_95 >= $now) {
+      $attr = 'icus';
+      $reason = 'gatt';
+      push @desc, "$last_domestic_copyright >= 1996 and $pub_year + 95 ($pub_plus_95) >= $now";
+    }
+    else {
+      $attr = 'pd';
+      $reason = ($is_pub)? 'exp':'add';
```

Comment:
Again, consider explicitly extracting separate behavior for author death date vs. corporate author with date of publication. It may result in more duplication but will probably make the behavior clearer; then we can decide if that duplication is meaningful or incidental.

Reply:
I think this would muddy the waters, as the ADD/pub difference doesn't play a role in the date logic. It only controls what rights reasons get used. I think the other changes I'm making should make the duplication unnecessary.